TAMUNA

Crabzmatic opened the pull request:
Implementation for https://github.com/adap/flower/issues/2033.

Crabzmatic commented:
Hi @jafermarq, @gsmalinovsky and me made a first draft of the pull request for TAMUNA algorithm. Could we discuss this with you in the near future?

Crabzmatic commented:
Hi @jafermarq, PR is ready for review. The `format-baseline.sh` script doesn't report any errors.

jafermarq commented on baselines/tamuna/pyproject.toml:
```diff
@@ -0,0 +1,144 @@
+
+[tool.poetry.dependencies]
+python = ">=3.8.15, <3.12.0"  # don't change this
+flwr = "1.3.0" # don't change this
+ray = "1.11.1" # don't change this
```
Let's update this to the latest version of Flower?, you just need to replace those lines with:
`flwr = { extras = ["simulation"], version = "1.5.0" }`

Crabzmatic replied:
done

jafermarq commented on baselines/tamuna/pyproject.toml:
```diff
@@ -0,0 +1,144 @@
+ray = "1.11.1" # don't change this
+hydra-core = "1.3.2" # don't change this
+matplotlib = "3.7.1"
+torch = {version = "2.0.1+cu118", source = "pytorch"}
+torchvision = {version = "0.15.2+cu118", source = "pytorch"}
+torchaudio = {version = "2.0.2+cu118", source = "pytorch"}
```
At least on my machine this was causing the download of these packages for all platforms (not just linux, but also windows... taking really a long time).  Did you also experience this? Would it be ok to replace these lines with the following?
```
torch = { url = "https://download.pytorch.org/whl/cu118/torch-2.0.1%2Bcu118-cp39-cp39-linux_x86_64.whl"}
torchvision = { url = "https://download.pytorch.org/whl/cu118/torchvision-0.15.2%2Bcu118-cp39-cp39-linux_x86_64.whl"}
torchaudio = { url = "https://download.pytorch.org/whl/cu118/torchaudio-2.0.2%2Bcu118-cp39-cp39-linux_x86_64.whl"}
```
And then remove the `pytorch` source a few lines below?

Crabzmatic replied:
This would work nice for Linux, but Windows packages would not be available and this would not work on Windows, right? If you plan to target mainly Linux, then this would be okay.

Crabzmatic replied:
done

jafermarq commented on baselines/tamuna/README.md:
````diff
@@ -0,0 +1,105 @@
+
+```bash
+# Create new conda environment with Python 3.9
+conda create -n tamuna python=3.9
+
+# Activate the newly created virtual environment
+conda activate tamuna
+
+# Install poetry using pip
+pip install poetry
+
+# In some cases there are issues when installing the poetry environment if chardet package is not previously installed
+pip install chardet
````
I'm not sure if it's good practice to create a poetry env inside a conda env (which might explain some of the issues you mention here). How about replacing these instructions with:
```
# set local python version via pyenv
pyenv local 3.9.17
# then fix that for poetry
poetry env use 3.9.17
# then install poetry env
poetry install
```

Crabzmatic replied:
Pyenv doesn't officially support Windows (except WSL), again this is not an issue if Linux has priority. Do we need Windows support?

jafermarq replied:
true. let's focus on ubuntu. We can add windows support later on (likely after _Summer of Reproducibility_)

Crabzmatic replied:
done

jafermarq commented on baselines/tamuna/README.md:
```diff
@@ -0,0 +1,105 @@
+
+## Expected Results
+
+The resulting directory in `./outputs/` should contain (among other things) `communication_complexity.png` and `loss_accuracy.png`.
```
I tested your code and the plots were created perfectly. Could you update the matplotlib code so a title for the figure is included (and maybe adding grid lines too?)

Crabzmatic replied:
I included the titles and added the grid lines.

jafermarq commented on baselines/tamuna/tamuna/conf/base.yaml:
```diff
@@ -0,0 +1,35 @@
+  client_device: 'cuda:0'
+  client_resources:
+    num_cpus: 4
+    num_gpus: 1
```
Since your model is fairly small, i'm wondering if it would be better to use a smaller `num_gpus`. For instance setting it to 0.2 would allow for 5 concurrent clients on a GPU (total <5GB of VRAM usage when running the exp.)

Crabzmatic replied:
I changed the resources to 2 cpus and 0.2 gpu.

jafermarq commented on baselines/tamuna/README.md:
````diff
@@ -0,0 +1,105 @@
+
+# Running time for default experimental setup is less than 14min on Intel Core i5-12400F and Nvidia GeForce RTX 3060 Ti.
+```
+
````
Could you also indicate to run in CPU-only mode?

Crabzmatic replied:
I added a comment about cpu-only and also added a cpu-only yaml config for reference.

jafermarq reviewed (commented):
Hey @Crabzmatic , I did an initial pass through your baseline. I ran the code making the changes i mention below (i.e. updating to flower 1.5 and fixing the torch/torchvision/torchaudio) version and it run perfectly.  Please find some comments below with suggestions to either add some changes or include some extra bits of info in the readme.

jafermarq commented on baselines/tamuna/README.md:
```diff
@@ -0,0 +1,107 @@
+
+> Note: If you use this baseline in your work, please remember to cite the original authors of the paper as well as the Flower paper.
+
+****Paper:**** https://arxiv.org/abs/2302.09832
```
```suggestion
****Paper:**** [arxiv.org/abs/2302.09832](https://arxiv.org/abs/2302.098320)
```

jafermarq commented on baselines/tamuna/README.md:
```diff
@@ -0,0 +1,107 @@
+
+****Hardware Setup:**** By default, the experiments expect at least one gpu, but this can be changed to cpu only by specifying client and server devices. Default setup less than 5 GB of dedicated GPU memory.
+
+****Contributors:**** Ivan Agarský @Crabzmatic, Grigory Malinovsky @gsmalinovsky
```
update with proper link to github profile? or is this linking to twitter (or something else?)

Crabzmatic replied:
added github links

jafermarq reviewed (commented):
I added a couple of extra minor comments. In addition to those could you:
* rename the `docs` directory you use for the images in the readme as `_static`, this will make it easier to integrate parts of your readme into the Flower docs.
* Since we created the `baseline_template` directory and its content we have made some changes to its `pyproject.toml`, most notably raising the `mypy` version. Could you update your `pyproject.toml` with  the new content in https://github.com/adap/flower/blob/main/baselines/baseline_template/pyproject.toml?

Crabzmatic commented:
> I added a couple of extra minor comments. In addition to those could you:
> 
> * rename the `docs` directory you use for the images in the readme as `_static`, this will make it easier to integrate parts of your readme into the Flower docs.
> * Since we created the `baseline_template` directory and its content we have made some changes to its `pyproject.toml`, most notably raising the `mypy` version. Could you update your `pyproject.toml` with  the new content in https://github.com/adap/flower/blob/main/baselines/baseline_template/pyproject.toml?

done

jafermarq commented on baselines/tamuna/README.md:
````diff
@@ -0,0 +1,98 @@
+
+```bash
+# set local python version via pyenv
+pyenv local 3.9.17
+# then fix that for poetry
+poetry env use 3.9.17
````
```suggestion
pyenv local 3.10.6
# then fix that for poetry
poetry env use 3.10.6
```

jafermarq replied:
We very recently set baselines to use >py3.10.0 (this is needed when running the test for merge) -- there is another suggestion about the `pyproject.toml` regarding this.

jafermarq commented on baselines/tamuna/pyproject.toml:
```diff
@@ -0,0 +1,141 @@
+]
+
+[tool.poetry.dependencies]
+python = ">=3.8.15, <3.12.0"
```
```suggestion
python = ">=3.10.0, <3.11.0"
```

jafermarq commented on baselines/tamuna/pyproject.toml:
```diff
@@ -0,0 +1,141 @@
+flwr = { extras = ["simulation"], version = "1.5.0" }
+hydra-core = "1.3.2"
+matplotlib = "3.7.1"
+torch = { url = "https://download.pytorch.org/whl/cu118/torch-2.0.1%2Bcu118-cp39-cp39-linux_x86_64.whl"}
+torchvision = { url = "https://download.pytorch.org/whl/cu118/torchvision-0.15.2%2Bcu118-cp39-cp39-linux_x86_64.whl"}
+torchaudio = { url = "https://download.pytorch.org/whl/cu118/torchaudio-2.0.2%2Bcu118-cp39-cp39-linux_x86_64.whl"}
```
```suggestion
torch = { url = "https://download.pytorch.org/whl/cu118/torch-2.0.1%2Bcu118-cp310-cp310-linux_x86_64.whl"}
torchvision = { url = "https://download.pytorch.org/whl/cu118/torchvision-0.15.2%2Bcu118-cp310-cp310-linux_x86_64.whl"}
torchaudio = { url = "https://download.pytorch.org/whl/cu118/torchaudio-2.0.2%2Bcu118-cp310-cp310-linux_x86_64.whl"}
```

jafermarq reviewed (commented):
Thanks for incorporating those changes.

I have done another pass and included some small comments. In addition to those:
* Are you planning to also include results from Figure 2&3 as discussed in #2033? 
* I see the formatting script passes perfectly! Great!
* The tests script however (i.e. `./dev/test-baseline.sh`) finds a few errors (mostly easy to fix)

Crabzmatic commented:
> Thanks for incorporating those changes.
> 
> I have done another pass and included some small comments. In addition to those:
> 
> * Are you planning to also include results from Figure 2&3 as discussed in [TAMUNA #2033](https://github.com/adap/flower/issues/2033)?
> * I see the formatting script passes perfectly! Great!
> * The tests script however (i.e. `./dev/test-baseline.sh`) finds a few errors (mostly easy to fix)

- Neural network experiments on MNIST, which were not present in the paper, seem more logical to me as baselines in Flower. In my opinion, it makes sense to not include original logistical regression experiments on w8a and real-sim datasets in Flower. It would make more sense if Scaffnew, Scaffold and 5GCS were already implemented, but as of now neither of these algorithms are. I would be okay with adding them later once these are implemented. Let me know what you think about this.
- I fixed most of the `mypy` errors, one remains: 

```
tamuna/strategy.py:152: error: Argument 2 to "FitIns" has incompatible type "dict[str, int]"; expected "dict[str, bool | bytes | float | int | str]"  [arg-type]
tamuna/strategy.py:152: note: "Dict" is invariant -- see https://mypy.readthedocs.io/en/stable/common_issues.html#variance
tamuna/strategy.py:152: note: Consider using "Mapping" instead, which is covariant in the value type
```

If I understand this correctly, problem can be solved by changing type hint in the definition of FitIns (see https://stackoverflow.com/questions/73603289/why-doesnt-parameter-type-dictstr-unionstr-int-accept-value-of-type-di)

as for the `pylint` warnings, I fixed some that made more sense, these remain:

```
************* Module tamuna.client
tamuna/client.py:19:0: R0902: Too many instance attributes (9/7) (too-many-instance-attributes)
tamuna/client.py:24:4: R0913: Too many arguments (6/5) (too-many-arguments)
tamuna/client.py:133:8: C0103: Attribute name "lr" doesn't conform to snake_case naming style (invalid-name)
tamuna/client.py:122:4: R0913: Too many arguments (6/5) (too-many-arguments)
tamuna/client.py:201:0: C0103: Argument name "lr" doesn't conform to snake_case naming style (invalid-name)
************* Module tamuna.dataset_preparation
tamuna/dataset_preparation.py:103:0: R0913: Too many arguments (6/5) (too-many-arguments)
tamuna/dataset_preparation.py:103:0: R0914: Too many local variables (22/15) (too-many-locals)
tamuna/dataset_preparation.py:178:8: E1121: Too many positional arguments for method call (too-many-function-args)
************* Module tamuna.models
tamuna/models.py:72:0: C0103: Argument name "lr" doesn't conform to snake_case naming style (invalid-name)
tamuna/models.py:72:0: R0913: Too many arguments (10/5) (too-many-arguments)
tamuna/models.py:140:0: C0103: Argument name "lr" doesn't conform to snake_case naming style (invalid-name)
tamuna/models.py:174:0: C0103: Argument name "lr" doesn't conform to snake_case naming style (invalid-name)
tamuna/models.py:174:0: R0913: Too many arguments (6/5) (too-many-arguments)
tamuna/models.py:229:0: C0103: Argument name "lr" doesn't conform to snake_case naming style (invalid-name)
tamuna/models.py:229:0: R0913: Too many arguments (6/5) (too-many-arguments)
************* Module tamuna.strategy
tamuna/strategy.py:47:0: R0902: Too many instance attributes (8/7) (too-many-instance-attributes)
tamuna/strategy.py:50:4: R0913: Too many arguments (6/5) (too-many-arguments)
```

Let me know if I need to fix these too.

jafermarq commented:
For `mypy` issue you could do:
```
local_epochs: Scalar = self.epochs_per_round
config = {"epochs": local_epochs}
```
after importing `Scalar` from `flwr.common`. I can't tell immediately why this error is being raised...

For the `pylint` you can add use the `ignore` syntax to skip all those saying `too-many-...`. Those for `invalid-name`, you can for instance replace `lr` with `l_r`.

Crabzmatic commented:
> For `mypy` issue you could do:
> 
> ```
> local_epochs: Scalar = self.epochs_per_round
> config = {"epochs": local_epochs}
> ```
> 
> after importing `Scalar` from `flwr.common`. I can't tell immediately why this error is being raised...
> 
> For the `pylint` you can add use the `ignore` syntax to skip all those saying `too-many-...`. Those for `invalid-name`, you can for instance replace `lr` with `l_r`.

Done, `test-baseline.sh` script report no errors/warnings now.

jafermarq commented on baselines/tamuna/README.md:
```diff
@@ -0,0 +1,98 @@
+
+The resulting directory in `./outputs/` should contain (among other things) `communication_complexity.png` and `loss_accuracy.png`.
+
+![Communication complexity](_static/communication_complexity.png)
+![Loss and accuracy](_static/loss_accuracy.png)
```
how about putting these side-by-side? 
```
<img src="_static/communication_complexity.png" width="500"/> <img src="_static/loss_accuracy.png" width="500"/> 
```

Crabzmatic replied:
Done, looks better now

jafermarq commented on baselines/tamuna/tamuna/main.py:
```diff
@@ -0,0 +1,203 @@
+                "num_cpus": cfg.client.client_resources.num_cpus,
+                "num_gpus": cfg.client.client_resources.num_gpus,
+            },
+            strategy=CentralizedFedAvgStrategy(
```
I wonder if it would be better not to define your own strategy in this case and instead directly use the built-in `FedAvg`. You can of course still override the `evaluate_fn` and the config sent to clients during `configure_fit` list shown below. I think it's preferable to work like this.

```python

# somewhere in your main you can define this
def get_fit_config(is_tamuna: bool):
    def fit_config(server_round: int) -> Dict[str, Scalar]:
        """Return a configuration with static batch size and (local) epochs."""
        if is_tamuna:
            # prepare config
        else:
            local_epochs: Scalar = cfg.server.clients_per_round
            config = {
                "epochs": local_epochs,  # Number of local epochs done by clients
            }
        return config
    return fit_config

strategy = fl.server.strategy.FedAvg(
    ... # add all other relevant arguments,
    evaluate_fn=evaluate_fn,
    on_fit_config_fn=get_fit_config(False),
    )
```

jafermarq replied:
I left some other comments in the `strategy.py`

Crabzmatic replied:
Ok, I removed my custom class for FedAvg and wrote it using built-in FedAvg.

jafermarq commented on baselines/tamuna/tamuna/strategy.py:
```diff
@@ -0,0 +1,180 @@
+    return pattern
+
+
+class TamunaStrategy(Strategy):
```
Following the logic in my comment on the `readme.md`, you could also make the `Tamuna` strategy, to inherit from `FedAvg` and only override those methods that are different from plain `FedAvg`.  For instance `initialize_parameters()` is never used...

Crabzmatic replied:
Are you referring to `initialize_parameters()` of `TamunaStrategy`? This method is mainly used to initialize the compression pattern for clients and is called and executed once at the beginning. I think I can't reuse any of the `FedAvg` methods in `Tamuna` unfortunately. One of the reasons is that compression needs to be called after training on clients (cannot be called at evaluate). I could theoretically use evaluate for clients to do the compression, but that seems very forced and not clean.

Crabzmatic replied:
`configure_fit` needs to be overridden to add shuffling of the compression pattern
`aggregate_fit` different way of aggregating, needs to be overriden
`initialize_parameters` is different, 
`configure_evaluate` and `aggregate_evaluate` are not used in `Tamuna`

I see two options: extend `FedAvg` and override all of the methods or write `Tamuna` as a separate strategy (extending only base `Strategy`). I think second is better in this case.

jafermarq replied:
ok. let's leave the `Tamuna` strategy as is.

jafermarq reviewed (commented):
Hi @Crabzmatic,

I have left a couple of comments. The main one is regarding how you define strategies: I think it would be better if you build them on top of the native `FedAvg` (there will be less code and will make them easier to maintain in the long term). Let me know if you have any doubts on how to do this. At least your implementation of `FedAvg` should be fairly straightforward to implement using the native one and passing the `on_fit_config` and `evaluate_fn` arguments to it. Let me know what you think!

I think after that we can start wrapping this up and start the process of merging your baseline with Flower's `main` branch 🙌

jafermarq commented on baselines/tamuna/pyproject.toml:
```diff
@@ -0,0 +1,141 @@
+version = "1.0.0"
+description = "Flower Baselines"
+license = "Apache-2.0"
+authors = ["The Flower Authors <hello@flower.dev>"]
```
I thought you might want to add your details to this `authors` field. Similarly, please update the `description` field above with just "tamuna" or maybe the full title of the Tamuna paper?

Crabzmatic replied:
added

jafermarq reviewed (commented):
Hi @Crabzmatic,

Just small comment. I think this looks ready to merge.

jafermarq commented on baselines/tamuna/tamuna/client.py:
```diff
@@ -0,0 +1,242 @@
+    trainloaders: List[DataLoader],
+    learning_rate: float,
+    model: DictConfig,
+    client_device: str,
```
```suggestion
```

Crabzmatic replied:
removed `client_device` in all places

jafermarq commented on baselines/tamuna/tamuna/client.py:
```diff
@@ -0,0 +1,242 @@
+        The learning rate for the SGD optimizer for clients.
+    model: DictConfig
+        Architecture of the model being instantiated
+    client_device: str
+        Device to use for client training (cpu, cuda)
```
```suggestion
```

jafermarq commented on baselines/tamuna/tamuna/client.py:
```diff
@@ -0,0 +1,242 @@
+    def fedavg_client_fn(cid: str) -> FedAvgClient:
+        """Create a FedAvg client."""
+        # Load model
+        device = torch.device(device=client_device)
```
```suggestion
        device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
```

jafermarq commented on baselines/tamuna/tamuna/conf/base-cpu.yaml:
```diff
@@ -0,0 +1,35 @@
+  downlink_factor: 1
+
+client:
+  client_device: 'cpu'
```
```suggestion
```

jafermarq commented on baselines/tamuna/tamuna/client.py:
```diff
@@ -0,0 +1,242 @@
+    def tamuna_client_fn(cid: str) -> TamunaClient:
+        """Create a Tamuna client."""
+        # Load model
+        device = torch.device(device=client_device)
```
```suggestion
        device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
```

jafermarq commented on baselines/tamuna/tamuna/client.py:
```diff
@@ -0,0 +1,242 @@
+    trainloaders: List[DataLoader],
+    learning_rate: float,
+    model: DictConfig,
+    client_device: str,
```
```suggestion
```

jafermarq commented on baselines/tamuna/tamuna/client.py:
```diff
@@ -0,0 +1,242 @@
+        The learning rate for Tamuna optimizer for clients.
+    model: DictConfig
+        Architecture of the model being instantiated
+    client_device: str
+        Device to use for client training (cpu, cuda)
```
```suggestion
```

jafermarq commented on baselines/tamuna/tamuna/main.py:
```diff
@@ -0,0 +1,213 @@
+        trainloaders=trainloaders,
+        learning_rate=cfg.client.learning_rate,
+        model=cfg.model,
+        client_device=cfg.client.client_device,
```
```suggestion
```

jafermarq commented on baselines/tamuna/tamuna/main.py:
```diff
@@ -0,0 +1,213 @@
+        trainloaders=trainloaders,
+        learning_rate=cfg.client.learning_rate,
+        model=cfg.model,
+        client_device=cfg.client.client_device,
```
```suggestion
```

jafermarq commented on baselines/tamuna/pyproject.toml:
```diff
@@ -0,0 +1,141 @@
+version = "1.0.0"
+description = "TAMUNA: Doubly Accelerated Federated Learning with Local Training, Compression, and Partial Participation"
+license = "Apache-2.0"
+authors = ["Ivan Agarský <agarskyivan@gmail.com>, Grigory Malinovsky"]
```
The tests fail (on the CI) because the expected format is "name <email>". and comma separated. Eg. `["Ivan Agarský <agarskyivan@gmail.com>", "Grigory Malinovsky <email@gmail.com>"]`

jafermarq reviewed (commented):
Hi @Crabzmatic,

Your baseline is ready to be merged. Just a couple of suggestions before going ahead:
* Let's not make use of `client_device` passed from the config to the functions that return client objects. Instead, you can let it be "discovered" (as suggested) based on the `client_resources` passed to `start_simulation`. It works fine. All but one suggestion below refer to this. you might need to re-run the formatting checks afterwards.
* Small change is needed to the `authors` field in the `pyproject.toml`.


Ping me once done, then we can merge `Tamuna`.
